[#3159] `AvroSerializer` - Documentation, code sample, and auto-config fix (#3275)
AxonFramework/AxonFramework: docs/old-reference-guide/modules/ROOT/pages/serialization.adoc

It produces a more compact serialized form, while requiring those classes to stick to Jackson's conventions (or configuration).
The compact format makes it ideal for events, commands, and queries, as it minimizes the storage space and package size.

> The third option provided by the AxonFramework is the `AvroServializer`. The `Serializer` uses https://avro.apache.org/[Apache Avro] to serialize objects into and deserialize from https://avro.apache.org/docs/1.12.0/specification/#binary-encoding[binary single-object-encoded format]. The format is very compact, is based on the https://avro.apache.org/docs/1.12.0/specification/[Avro Schema Specification] and is ideal for messaging (serialization of commands, events and queries) and event storage. This serializer is designed to serve as payload `event`/`message` serializer and requires a `delegate` serializer to serialize everything else (here the best choice is `JacksonSerializer`). Since the entire serialization process is based on Apache Avro schema, a schema store is required. The framework provides a local `SchemaStore` which can be pre-filled manually or uses SpringBoot class path scanning to detect  schemas located on the class path.

Reviewer: ```suggestion
The third option provided by the AxonFramework is the `AvroServializer`. The `Serializer` uses https://avro.apache.org/[Apache Avro] to serialize objects into and deserialize from https://avro.apache.org/docs/1.12.0/specification/#binary-encoding[binary single-object-encoded format]. The format is very compact, is based on the https://avro.apache.org/docs/1.12.0/specification/[Avro Schema Specification] and is ideal for messaging (serialization of commands, events and queries) and event storage. This serializer is designed to serve as payload `event`/`message` serializer and requires a `delegate` serializer to serialize everything else (here the best choice is `JacksonSerializer`). Since the entire serialization process is based on Apache Avro schema, a schema store is required. The framework provides a local `SchemaStore` which can be pre-filled manually or uses Spring Boot class path scanning to detect schemas located on the class path.
```

}
----

===== Spring Boot - configuration class using XStream/AvroSerializer

[source,java]
----
@Configuration
public class SerializerConfiguration {

   // By default, we want the XStreamSerializer
   @Bean
   public Serializer defaultSerializer() {
      // Set the secure types on the xStream instance
      XStream xStream = new XStream();
      return XStreamSerializer.builder()
                              .xStream(xStream)
                              .build();
   }

   // But for all our messages we'd prefer the AvroSerializer due to Avro's smaller format
   @Bean
   @Qualifier("messageSerializer")
   public Serializer messageSerializer(SchemaStore schemaStore) {
      return AvroSerializer.builder()
                           .delegateSerializer(JacksonSerializer.defaultSerializer())
                           .schemaStore(schemaStore)
                           .build();
   }
}
----

===== Spring Boot - properties file

[source,text]
----

> # Possible values for these keys are `default`, `xstream`, `java`, and `jackson`. `avro` should be used for `events` and `messages` only.

Reviewer: ```suggestion
# Possible values for these keys are `default`, `xstream`, `java`, and `jackson`. `avro` can be used for `events` and `messages` only.
```

[source,yaml]
----

> # Possible values for these keys are `default`, `xstream`, `java`, and `jackson`. `avro` should be used for `events` and `messages` only.

Reviewer: ```suggestion
# Possible values for these keys are `default`, `xstream`, `java`, and `jackson`. `avro` can be used for `events` and `messages` only.
```

The `DataInputStream` and `DataOutputStream` allow you to easily write the contents of the events to an output stream.
The `ByteArrayOutputStream` and `ByteArrayInputStream` allow writing to and reading from byte arrays.

=== AvroSerializer

Avro Serializer requires a `SchemaStore` to operate, which must be able to resolve schemas of all messages being processed. By default, you can use default `org.apache.avro.message.SchemaStore.Cache` and add Avro Schema to it manually. If you are using

> SpringBoot, the autoconfiguration will create a `SchemaStore` instance for you. This Schema Store can be pre-filled with Avro Schemas extracted from Avro `SpecificRecordBase` generated by `avro-maven-plugin` located on your classpath. For this purpose,

Reviewer: ```suggestion
Spring Boot, the auto-configuration will create a `SchemaStore` instance for you. This Schema Store can be pre-filled with Avro Schemas extracted from Avro `SpecificRecordBase` instances. These are in turn generated by the `avro-maven-plugin` that should then be located on your classpath. For this purpose,
```

a special annotation `@AvroSchemaScan` should be put to any SpringBoot configuration class (similar to SpringData `@EntityScan`) in a package hierarchy including the `SpecificRecordBase` classes.

> In case you want to use a central Schema Registry in your application landscape, you might want to provide your own `SchemaStore`
> implementation, accessing this remote registry. Just implement the corresponding interface on your own and wire it into the configuration of the `AxonSerrializer`, but consider to implement a caching strategy, since the `AvroSerializer` will look the Avro Schema by its fingerprint on any serialization / deserialization operation, making fast and local in-memory resolution important to keep up the performance.

Reviewer: ```suggestion
In case you want to use a central Schema Registry in your application landscape, you should provide your own `SchemaStore`
implementation, accessing this remote registry. Just implement the corresponding interface and wire it into the configuration of the `AvroSerrializer`. When doing this, do consider to implement a caching strategy in your implementation. The `AvroSerializer` will check the Avro Schemas by their fingerprints on any serialization / deserialization operation, making fast and local in-memory resolution important to keep up the performance.
```

> The `AxonSerializer` is extendable and delegates to `AvroSerializerStrategy` to support different flavors of source classes. The default strategy (enabled by default) is the `SpecificRecordBaseSerializerStrategy` which is capable to work with Java classes generated by `avro-maven-plugin`. If you use different representation of Avro classes in your project, you might want to supply a different `AvroSerializerStrategy` in order to copy with specifics of your implementation. For example, to support KotlinX serialization of data classes, implemented by https://github.com/avro-kotlin/avro4k[avro4k], there is a community project, delivering the corresponding `AvroSerializerStrategy`.

Reviewer: ```suggestion
The `AvroSerializer` is extendable and delegates to `AvroSerializerStrategy` to support different flavors of source classes. The default strategy is the `SpecificRecordBaseSerializerStrategy` that is capable to work with Java classes generated by `avro-maven-plugin`. If you use different representation of Avro classes in your project, you might want to supply a different `AvroSerializerStrategy` in order to deal with the specifics of your implementation. For example, to support KotlinX serialization of data classes, implemented by https://github.com/avro-kotlin/avro4k[avro4k], there is a community project, delivering the corresponding `AvroSerializerStrategy`.
```
